python/tests/core/test_delta_engine.py

```diff
@@ -0,0 +1,296 @@
+        _patch_client(mocker, is_external=True, project_name="prj")
+        _patch_apis(mocker, lb_domain="dn.example.com", username=None)
+
+        fg = _make_fg("hopsfs://nn:8020/projects/p1")
+        with pytest.raises(FeatureStoreException) as e:
+            DeltaEngine(
+                feature_store_id=1,
+                feature_store_name="fs",
+                feature_group=fg,
+                spark_session=None,
+                spark_context=None,
+            )
+        assert "Cannot get user name" in str(e.value)
+
+    def test_get_delta_rs_location_internal(self, mocker):
+        _patch_client(mocker, is_external=False)
+        _patch_apis(mocker, lb_domain="nn.example.com", username="user1")
+
+        fg = _make_fg("hopsfs://nn:8020/projects/p1")
+        engine = DeltaEngine(1, "fs", fg, None, None)
+        loc = engine._get_delta_rs_location()
+        assert loc == "hdfs://nn:8020/projects/p1"
+
+    def test_get_delta_rs_location_external_success(self, mocker):
+        _patch_client(mocker, is_external=True)
+        _patch_apis(mocker, lb_domain="nn.lb.example.com", username="user1")
+
+        fg = _make_fg("hopsfs://nn:8020/projects/p1")
+        engine = DeltaEngine(1, "fs", fg, None, None)
+        loc = engine._get_delta_rs_location()
+        assert loc == "hdfs://nn.lb.example.com:8020/projects/p1"
+
+    def test_get_delta_rs_location_external_error(self, mocker):
+        _patch_client(mocker, is_external=True)
+        # Allow _setup_delta_rs to succeed (needs datanode), but fail on namenode during location
+        _patch_apis(
+            mocker,
+            username="user1",
+            datanode_lb="dn.example.com",
+            namenode_lb=None,
+        )
+
+        fg = _make_fg("hopsfs://nn:8020/projects/p1")
+        engine = DeltaEngine(1, "fs", fg, None, None)
+        with pytest.raises(FeatureStoreException) as e:
+            engine._get_delta_rs_location()
+        assert "namenode load balancer" in str(e.value)
+
+    def test_setup_delta_read_opts_variants(self, mocker):
+        _patch_client(mocker, is_external=False)
+        fg = _make_fg("hopsfs://nn:8020/p")
+        engine = DeltaEngine(1, "fs", fg, None, None)
+
+        alias = mock.Mock()
+        alias.left_feature_group_end_timestamp = None
+        alias.left_feature_group_start_timestamp = None
+        assert engine._setup_delta_read_opts(alias, None) == {}
+
+        alias.left_feature_group_end_timestamp = 1234567890
+        alias.left_feature_group_start_timestamp = None
+        mocker.patch("hsfs.util.get_delta_datestr_from_timestamp", return_value="t")
+        assert engine._setup_delta_read_opts(alias, None) == {
+            engine.DELTA_QUERY_TIME_TRAVEL_AS_OF_INSTANT: "t"
+        }
+
+        # merging read_options
+        opts = engine._setup_delta_read_opts(alias, {"k": "v"})
+        assert opts[engine.DELTA_QUERY_TIME_TRAVEL_AS_OF_INSTANT] == "t"
+        assert opts["k"] == "v"
```

Comment:
tests like these will be hard to understand/maintain as it is testing multiple things

each case should have its own test.